polling_stations/apps/addressbase/models.py

```diff
@@ -18,6 +27,59 @@ def points_for_postcode(self, postcode):
         qs = qs.values_list("location", flat=True)
         return list(qs)
 
+    def get_uprn(self, uprn):
+        cursor = connection.cursor()
+        cursor.execute(
+            """
+                SELECT a.address, a.postcode, u.lad, u.uprn, u.polling_station_id, a.location
+                FROM addressbase_address a
+                    JOIN addressbase_uprntocouncil u
+                    ON a.uprn = u.uprn
+                WHERE a.uprn=%s
+            """,
```

Comment:
First mini-review of many

@symroe actually fixed this in https://github.com/DemocracyClub/uk-geo-utils/pull/16 but the package version on PyPI is still 0.9.0: https://pypi.org/project/uk-geo-utils/
If you build the package and publish it to pypi, then you can upgrade to 0.10.0 and you'll be able to do this.

Reply:
I never updated the package, sorry. I've just published `0.10.0`

Reply:
I'm going to write quite a lot about this one snippet, but I think this is quite an central bit of logic so it is worth getting the abstraction right here.

First thing is that https://pypi.org/project/uk-geo-utils/0.10.1/ would allow you to get rid of the raw SQL here so that is probably woth doing.

Maybe that tidies things up a bit, but secondly, this `get_uprn` function is really wierd. It is called `get_uprn` but we have to already have a UPRN to call it and then it returns.. not a UPRN. Also its a function on the Address collection, but its really about the properties of a single Address object. It seems like what we're really trying to do here is compensate for the fact that we would like to get an Address object, but some of the fields we want an address object to have don't live in the address table. A more normal way to deal with this might be to augment the Address model with some extra properties e.g:

```py
class Address(AbstractAddress):
    objects = AddressManager()

    @property
    def polling_station_id(self):
        # get polling_station_id off the uprntocouncil table
        # return it

    @property
    def council_id(self):
        # get lad off the uprntocouncil table
        # return it
```

(or possibly `.council_id` should actually be `.council` and return a council object?)

Then `Address.object.get(uprn=12345)` gets us an object with all the things we need.

Just thinking back to https://github.com/DemocracyClub/UK-Polling-Stations/pull/2703#issuecomment-731599600 and whether to replace `PollingStation.objects.get_polling_station_by_id(internal_council_id, council_id)` with `PollingStation.objects.get_polling_station_by_uprn(uprn)`, maybe if the entry point for getting a polling station is _always_ going to be that we've got an `Address` record, maybe the right abstraction is actually to move that logic onto the address object?


```py
class Address(AbstractAddress):
    #...

    @property
    def polling_station(self):
        station = self.filter(
            internal_council_id=self.polling_station_id, council_id=self.council_id
        )
        if len(station) == 1:
            return station[0]
        else:
            return None
```

Does any of that seem like it might help clean things up a bit here? Have a think about it..

Reply:
### Joining to UprnToCouncil
This works nicely, it even helped me tidy a bunch of tests :smile: . The down side is that running `./manage.py import_uprn_council_lookup uprn-to-councils.csv` now takes ~25mins instead of ~6min. This isn't a huge issue though.
I haven't tackled joining `UprnToCouncil` to the `Councils` or `PollingStations`, as that broke things differently. 

### Properties
I really like using properties for the council and polling station ids/objects. It's a great way of simplifying the interface to this information. As I guess you anticipated it let's us get rid of the `get_uprn`/`uprns_for_postcode` methods on `AddressManager`. I've left in `uprns_for_postcode` as the dashboard views use it, and I'd rather punt fixing that down the road.

Another potential consequence is making the addressbase geocoder redundant. I'm not certain about this, but I think the only thing we're missing is postcode centroid, which I think could also be a property on the Address model? Guess that would simplify the geocoder code a bit, but again is not for this PR.